core/lib/zksync_core/src/sync_layer/batch_status_updater/mod.rs

```diff
@@ -0,0 +1,451 @@
+                }
+                Err(UpdaterError::Internal(err)) => return Err(err),
+            }
+
+            if status_changes.is_empty() {
+                tokio::time::sleep(self.sleep_interval).await;
+            } else {
+                self.apply_status_changes(&mut cursor, status_changes)
+                    .await?;
+            }
+        }
+    }
+
+    /// Goes through the already fetched batches trying to update their statuses.
+    /// Returns a collection of the status updates grouped by the operation type.
+    ///
+    /// Fetched changes are capped by the last locally applied batch number, so
+    /// it's safe to assume that every status change can safely be applied (no status
+    /// changes "from the future").
+    async fn get_status_changes(
+        &self,
+        status_changes: &mut StatusChanges,
+        mut cursor: UpdaterCursor,
+    ) -> Result<(), UpdaterError> {
+        let total_latency = EN_METRICS.update_batch_statuses.start();
+        let Some(last_sealed_batch) = self
+            .pool
+            .access_storage_tagged("sync_layer")
+            .await?
+            .blocks_dal()
+            .get_sealed_l1_batch_number()
+            .await?
+        else {
+            return Ok(()); // No L1 batches in the storage yet; do nothing.
+        };
+
+        let mut batch = cursor.last_executed_l1_batch.next();
+        // In this loop we try to progress on the batch statuses, utilizing the same request to the node to potentially
+        // update all three statuses (e.g. if the node is still syncing), but also skipping the gaps in the statuses
+        // (e.g. if the last executed batch is 10, but the last proven is 20, we don't need to check the batches 11-19).
+        while batch <= last_sealed_batch {
+            // While we may receive `None` for the `self.current_l1_batch`, it's OK: open batch is guaranteed to not
+            // be sent to L1.
+            let miniblock_number = self.client.resolve_l1_batch_to_miniblock(batch).await?;
+            let Some(miniblock_number) = miniblock_number else {
+                return Ok(());
+            };
+
+            let Some(batch_info) = self.client.block_details(miniblock_number).await? else {
+                // We cannot recover from an external API inconsistency.
+                let err = anyhow::anyhow!(
+                    "Node API is inconsistent: miniblock {miniblock_number} was reported to be a part of {batch} L1 batch, \
+                    but API has no information about this miniblock",
+                );
+                return Err(err.into());
+            };
+
+            for stage in [
+                AggregatedActionType::Commit,
+                AggregatedActionType::PublishProofOnchain,
+                AggregatedActionType::Execute,
+            ] {
+                cursor.update(status_changes, &batch_info, stage)?;
```

Comment:
a small refactor suggestion: for me the code would be relatively easier to read if this loop was moved inside cursor's update method